src/system/AutoFreePacketBuffer.h

```diff
@@ -0,0 +1,67 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include "SystemPacketBuffer.h"
+
+namespace chip {
+namespace System {
+
+/// Ensures that a System::PacketBuffer is freed when it goes out of scope.
+class AutoFreePacketBuffer
+{
+public:
+    AutoFreePacketBuffer() : mBuffer(nullptr) {}
+    explicit AutoFreePacketBuffer(PacketBuffer * p) : mBuffer(p) {}
+
+    AutoFreePacketBuffer(const AutoFreePacketBuffer &) = delete;
+    AutoFreePacketBuffer & operator=(const AutoFreePacketBuffer &) = delete;
+
+    ~AutoFreePacketBuffer() { Adopt(nullptr); }
+
+    inline PacketBuffer * operator->() { return mBuffer; }
+    inline PacketBuffer & operator*() { return *mBuffer; }
+
+    inline const PacketBuffer * operator->() const { return mBuffer; }
+    inline const PacketBuffer & operator*() const { return *mBuffer; }
+
+    inline void Adopt(PacketBuffer * v)
+    {
+        if (mBuffer != nullptr)
+        {
+            PacketBuffer::Free(mBuffer);
```

Comment:
I believe it is clearer when reading the code. WIthout it, I need to go through PacketBuffer::Free doc or implementation to ensure it is null safe.

Is the concern that more code is generated and this wastes flash?
The code I replaced all did 'if buffer !=null then free' ... it did not seem to spark concerns there.

Reply:
There are two concerns: amount of code generated (which is pretty minor from my point of view) and the confusion by someone reading the code and knows `PacketBuffer::Free` is null-safe about why we have an extraneous null-check.  And if someone thinks that, they might be tempted to optimize and remove the null-check.  But I don't think we should, because that will involve taking a lock before the null-check in at least some configurations at the moment.  So worth adding a comment so people don't remove the null-check, maybe.

Reply:
In the (not too) long run, practically nothing else will ever call `PacketBuffer::Free` anyway.